### PR TITLE
Node API and CLI

### LINKED ISSUE #137: Simplified JavaScript API
Would you be open to a JavaScript API that wraps the babel API similar to how the CLI tool wraps babel-cli? For non Babel users trying to use babili programmatically is a little more complicated than it could be, because you have to first learn a different tool that's used in a different way.

### LINKED ISSUE #418: `babili --version` shows babel-cli version instead of babili version
I was trying to check if my global `babili` installation was up to date, but:

```
> babili --version
6.22.2 (babel-core 6.22.1)
```

I expected it to show `0.0.11 (babel-core 6.22.1)` instead.

It seems like a pretty tough change with how the babili CLI wraps the babel CLI, unfortunately…

### LINKED ISSUE #429: Babili CLI
Currently babili-cli is an alias to babel-cli with babili-preset and there is no way to specify babili options via CLI.

So, a new CLI that takes in babili-preset options () and some of babel options - for example - sourcemap support.

Examples:

```sh
babili --mangle=false # or babili --no-mangle
babili --mangle.topLevel 
babili --mangle.blacklist=Foo,Bar,Baz

babili --unsafe false
babili --unsafe.guards false # for option group
babili --unsafe.typeConstructors.string false # for delegating options to plugins in a group
```

All the options in [Preset Options](https://github.com/babel/babili/tree/master/packages/babel-preset-babili#options) should be available from the CLI.

File API: (examples)

```sh
babili --unsafe=false /path/to/input.js # outputs to STDOUT
babili --unsafe=false --out /path/to/out/dir /path/to/input/dir/*.js # takes multiple arguments

babili --unsafe=false --out /path/to/out --source-map /path/to/input.js # outputs sourcemap

# Option to provide input sourcemap - 
# Can be --input-source-map. But need some way to do for multiple files 
# Or can just NOT have support for soucemaps for multiple files via CLI
```

Related: 
+ #418 
+ Simplified JS API - #137

### LINKED ISSUE #484: 2 extra line breaks are added
```sh
$ node --version
v7.2.0

$ babili --version
6.24.0 (babel-core 6.24.0)

$ touch test.js

# Empty input
$ cat test.js | wc -c
       0

# 2-bytes output
$ babili test.js | wc -c
       2
```

### PR DESCRIPTION
**Pull request by boopathi**
+ Fix #484 
+ Fix #429 
+ Fix #418 
+ Fix #137

### CODE REVIEW

**Comment by sylvesteraswin** [comment]
What do you think of using `commander` to create the cli?

**Comment by hzoo** [comment]
Hey @boopathi! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vnode - linux](https://travis-ci.org/babel/babili/jobs/225141309)

**Comment by hzoo** [comment]
Hey @boopathi! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vnode - linux](https://travis-ci.org/babel/babili/jobs/225211366)

**Comment by hzoo** [comment]
Hey @boopathi! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vnode - linux](https://travis-ci.org/babel/babili/jobs/225214507)

**Comment by hzoo** [comment]
Hey @boopathi! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vnode - linux](https://travis-ci.org/babel/babili/jobs/225218822)

**Comment by hzoo** [comment]
Hey @boopathi! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vnode - linux](https://travis-ci.org/babel/babili/jobs/225248266)

**Comment by hzoo** [comment]
Hey @vigneshshanmugam! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vnode - linux](https://travis-ci.org/babel/babili/jobs/225351968)

File: .eslintrc.js [hunk 1]
```diff
@@ -14,6 +14,7 @@ module.exports = {
   rules: {
     "linebreak-style": ["error", "unix"],
     "no-cond-assign": OFF,
-    "no-case-declarations": OFF
+    "no-case-declarations": OFF,
+    "no-console": ["error", { allow: ["error"] }]
```
Comment (boopathi) [inline .eslintrc.js]: why ?

Comment (vigneshshanmugam) [inline .eslintrc.js]: for `console.error` from fs errors.

**Comment by hzoo** [comment]
Hey @vigneshshanmugam! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vnode - linux](https://travis-ci.org/babel/babili/jobs/225987478)

File: packages/babili/src/cli.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    boolean: booleanOpts,
+    array: mangleArrayOpts,
+    default: Object.assign({}, arrayDefaults, booleanDefaults),
+    alias,
+    configuration: {
+      "dot-notation": false
+    }
+  });
+
+  const files = argv["_"];
```
Comment (vigneshshanmugam) [inline packages/babili/src/cli.js]: only validation i would do extra is to validate the files and check if they are present.. did we need to do anything extra? @boopathi

File: .eslintrc.js (hunk 1 above)
Comment (boopathi) [inline .eslintrc.js]: `process.stderr.write()`

File: packages/babili/src/cli.js (hunk 2 above)
Comment (boopathi) [inline packages/babili/src/cli.js]: `fs.readFile` will already throw if the file is either not present or not read-able. So, I'm not sure if that needs to be separate verified.

Comment (vigneshshanmugam) [inline packages/babili/src/cli.js]: Yeah and I did also check ignore the file if it isn't present https://github.com/babel/babili/blob/2b1b16ac05596e65ec77c56a1e3e1b7882991341/packages/babili/src/fs.js#L29

File: packages/babili/src/cli.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  return r;
+}
+
+function printHelpInfo() {
+  const msg = `
+  Usage: babili index.js [options]
+
+  Options:
+    --out-file, -o          Output to a specific file
+
```
Comment (vigneshshanmugam) [inline packages/babili/src/cli.js]: we should keep a minimal list for the help command.. thoughts? @boopathi

File: package.json [hunk 4]
```diff
@@ -27,12 +27,13 @@
...
     },
     "testEnvironment": "node",
+    "testPathIgnorePatterns": [
+      "/node_modules/",
+      "/fixtures/"
+    ],
     "roots": [
       "packages"
     ],
-    "transformIgnorePatterns": [
```
Comment (vigneshshanmugam) [inline package.json]: not required. default is node_modules

File: packages/babili/src/fs.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  const { fileOpts, options: babiliOpts } = detachOptions(options);
+  const { stdin, outFile } = fileOpts;
+  if (stdin) {
+    readStdin().then(input => {
+      let { code } = babili(input, babiliOpts);
+      // write to stdout if ouput file is not specified
+      if (outFile === void 0) {
+        process.stdout.write(code + "\n");
+      } else {
+        fs.writeFileSync(path.resolve(outFile), code, "utf-8");
```
Comment (vigneshshanmugam) [inline packages/babili/src/fs.js]: should be changed to outputFileSync

File: packages/babili/__tests__/cli-tests.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  it("should throw on all invalid options", () => {
+    expect(spawn("", "--foo", "--bar").stderr).toMatchSnapshot();
+  });
+
+  it("should read from stdin and o/p to stdout", () => {
+    const source = "let abcd = 10, bcdsa = 20";
+    expect(spawn(source, "--mangle.topLevel").stdout).toMatchSnapshot();
+  });
+
+  xit("should handle input file and --out-file option", () => {
```
Comment (vigneshshanmugam) [inline packages/babili/__tests__/cli-tests.js]: there is issue with validation options, need to fix it first.

**Pull request merged**